src/permutation.rs

```diff
@@ -404,6 +458,75 @@ impl Permutation {
             true => self.apply_slice_fwd(vec),
         }
     }
+
+    /// Apply a permutation to a slice of elements
+    ///
+    /// Given a slice of elements, this will permute the elements in place according
+    /// to this permutation.
+    ///
+    /// This method borrows `self` mutably to avoid allocations, but the permutation
+    /// will be unchanged after it returns.
+    ///
+    /// Note that unlike the other `apply_*` methods, this method *requires* the permutation
```

Comment:
Maybe I'm tired right now, but why can't we do in place application in reverse?

Reply:
In all honesty I don't remember :laughing: 
I *think* my thought was that we could just normalize the permutation in the appropriate direction since we're already borrowing `self` mutably, but that feels like an excessive side-effect in a function whose whole purpose is to be efficient.

There's also the chance that I'm missing the point and that we don't need to normalize at all to do an in-place application, but I'm not sure about that.

Reply:
I think you can do an in-place forward application, iterating through a cycle and swapping with the first element, rather than the previous element in the cycle, each time.

e.g. with a cycle 

`1 2 3 5 7`

We swap (1,2), (1,3), (1,5), (1,7).

Reply:
I gave it a try, I'm not 100% I got it right but simple tests seem to pass.